test/test_images/internal/knockdown/knockdown.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2019 Google LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package knockdown
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+)
+
+// Config is a common envconfig filled struct that includes the common options for knockdown
+// testing. It is expected that knockdown tests will embed this struct in their configuration
+// struct.
+// E.g.
+// type FooKnockdownConfig {
+//     knockdown.Config
+//     FooProperty string `envconfig:...`
+// }
+type Config struct {
+	Time time.Duration `envconfig:"TIME" required:"true"`
+}
+
+// cancel will cancel the main receiver's context. Calling this will cause Main() to become
+// unblocked and call os.Exit(0). In general, write{Successful,Failed}TerminationMessage should be
+// called first.
+var cancel func()
+
+// Main should be called by the process' main function. It will run the knockdown test. It will not
+// return. Instead, it will call os.Exit.
+func Main(config Config, kdr Receiver) {
+	client, err := cloudevents.NewDefaultClient()
+	if err != nil {
+		panic(err)
+	}
+
+	r := receiver{
+		kdr: kdr,
+	}
+
+	var ctx context.Context
+	ctx, cancel = context.WithCancel(context.Background())
+
+	timer := time.NewTimer(config.Time)
+	defer timer.Stop()
+	go func() {
+		<-timer.C
+		// Write the termination message if time out
+		fmt.Printf("Time out waiting for the knock down event(s).")
+		if err := r.writeFailedTerminationMessage(); err != nil {
+			fmt.Printf("Failed to write termination message, %v\n", err)
+		}
+		cancel()
+	}()
+
+	fmt.Printf("Waiting to receive event (timeout in %s)...", config.Time.String())
+
+	// Note, we assume that closing the context will gracefully shutdown the receiver. Having looked
+	// at the current implementation, this is true. But nothing guarantees it in the future.
+	if err := client.StartReceiver(ctx, r.Receive); err != nil {
+		log.Fatal(err)
+	}
+
+	os.Exit(0)
```

Comment:
I agree it is likely not needed. But because the correctness of the tests depends on it happening, I would prefer to keep it here.

Reply:
My concern with calling os.Exit is that it will skip defers. It is very easy for someone to add a defer to this function and not realize that it won't get called. It's also possible that the caller will want to do some additional cleanup or perform additional checks after Main exits.

Reply:
Switched. Although I don't think it is useful in this instance, because this is for fairly simple test code.